xla/service/collective_pipeliner.cc

```diff
@@ -718,6 +738,28 @@ int64_t WhileLoopAnalysis::GetDUSIndex(const HloInstruction* dus) const {
   return it->second;
 }
 
+void WhileLoopAnalysis::ExtractLoopInvariantOps() {
+  for (HloInstruction* inst :
+       while_->while_body()->MakeInstructionPostOrder()) {
+    if (inst->opcode() == HloOpcode::kConstant) {
```

Comment:
I think hlo_query has some util function for walking over instuctions of an op type.